From 16055afced4808df8f9800f1844d2540b29d9d04 Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Mon, 7 Aug 2017 13:36:23 +0800
Subject: [PATCH 175/184] pcie-octeon.c: add a function
 octeon_pcie_get_qlm_from_fdt()

This function doesn't exist in Cavium's SDK formatted patches, but SDK's
final image(OCTEON-SDK-3.1.2-568 .i386.rpm + sdk_3.1.2_update_p10 .tgz),
has it, port it from SDK's final image.

Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 arch/mips/pci/pcie-octeon.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/arch/mips/pci/pcie-octeon.c b/arch/mips/pci/pcie-octeon.c
index 2ab6242..b00cecf 100644
--- a/arch/mips/pci/pcie-octeon.c
+++ b/arch/mips/pci/pcie-octeon.c
@@ -12,6 +12,7 @@
 #include <linux/time.h>
 #include <linux/delay.h>
 #include <linux/module.h>
+#include <linux/of_address.h>
 
 #include <asm/octeon/octeon.h>
 #include <asm/octeon/cvmx-npei-defs.h>
@@ -532,6 +533,30 @@ static void octeon_pcie_interface_init(struct octeon_pcie_interface *iface, unsi
 	iface->pem = pem;
 }
 
+int octeon_pcie_get_qlm_from_fdt(int numa_node, int pcie_port)
+{
+	struct device_node	*np;
+	const int		*qlm;
+
+	for_each_compatible_node(np, NULL, "cavium,octeon-7890-pcie") {
+		const __be32		*reg;
+		u64			addr;
+
+		reg = of_get_property(np, "reg", NULL);
+		if (reg) {
+			addr = of_translate_address(np, reg);
+			if ((numa_node == (addr >> 36 & 0x3)) &&
+			    (pcie_port == (addr >> 24 & 0x3))) {
+				qlm = of_get_property(np, "qlm", NULL);
+				return qlm ? *qlm : -1;
+			}
+		}
+	}
+
+	return -1;
+}
+EXPORT_SYMBOL(octeon_pcie_get_qlm_from_fdt);
+
 static struct pci_ops octeon_dummy_ops = {
 	.read	= octeon_dummy_read_config,
 	.write	= octeon_dummy_write_config,
-- 
1.9.1

